server/the_commander.py

```diff
@@ -0,0 +1,173 @@
+            raw = query.first()
+
+            username = raw.username
+            password = raw.password
+            mfa_token = raw.mfa_token
+
+            self.logger.warning('New credentials submitted. Verifying..')
+            if not username:
+                self.logger.warning('No username submitted but someone posted a web response. Aborting')
+                return
+            else:
+                username = raw.username.strip()
+                self.logger.warning('Username: {0}'.format(username))
+
+            if not password:
+                self.logger.warning('No password submitted for {0}. Aborting'.format(username))
+                return
+
+            if not mfa_token:
+                if self.config['mfa_required']:
+                    self.logger.warning('MFA is required but no MFA submitted for {0}. Aborting'.format(username))
+                    return
+                mfa_token = ''
+            else:
+                mfa_token = raw.mfa_token.strip()
+
+            if len(username) > self.config['username_len']:
+                self.logger.warning('Username length is too long. Maximum is {0} but {1} was entered'.format(self.config['username_len'], len(username)))
+                return
+
+            if len(mfa_token) > self.config['mfa_token_len']:
+                self.logger.warning('MFA length is too long. Maximum is {0} but {1} was entered'.format(self.config['mfa_token_len'], len(mfa_token)))
+                return
+
+            if len(password) > self.config['password_len']:
+                self.logger.warning('Password length is too long. Maximum is {0} but {1} was entered'.format(self.config['password_len'], len(password)))
+                return
+
+            if '\\' in username:
+                if len(username.split()) > 2:
+                    self.logger.warning('Aborting due to too many backslashes in username: {0}'.format(username))
+                    return
+                username = username.split('\\')[-1]
+
+            illegal_chars = ['/', '\\', '[', ']', ':', ';', '|', '=', ',', '+', '*', '?', '<', '>', ' ', '&', '!', '~', '#', '%', '^', '(', ')', '{', '}' '`']
+            for illegal in illegal_chars:
+                if illegal in username:
+                    self.logger.warning('Aborting. Found illegal character in username: {0}'.format(illegal))
+                    return
+                if illegal in mfa_token:
+                    self.logger.warning('Aborting. Found illegal character in MFA: {0}'.format(illegal))
+                    return
+
+            username = re.escape(username)
+            password = re.escape(password)
+            mfa_token = re.escape(mfa_token)
+
+            # Execute command logic here
+            self.logger.warn('Command:\n{0}'.format(self.config['command'].format(username, '<REDACTED>', mfa_token)))
+            command = self.config['command'].format(username, password, mfa_token)
```

Comment:
Use kwargs here like `.format(username=username, password=password, mfa_token=mfa_token)`, unless you decide to go with Jinja.

Reply:
I had no idea you could use kwargs here. This is an excellent modification.